### PR TITLE
Feature/355 nested dragging

### PR DESCRIPTION
**Pull request by Nicholas-Westby**
Allow dragging of fieldsets between nesting barriers and to other Archetype properties: https://github.com/imulus/Archetype/issues/355

### CODE REVIEW

File: app/controllers/controller.js [hunk 1]
```diff
@@ -61,7 +62,36 @@ angular.module("umbraco").controller("Imulus.ArchetypeController", function ($sc
...
+            if (sourceScope && !sameScope) {
+
+                // Variables.
+                var targetFieldsets = targetScope.model.config.fieldsets;
+                var index = ui.item.sortable.index;
+                var properties = sourceScope.model.value.fieldsets[index].properties;
+                var canRemove = sourceScope.canRemove();
+                var canAdd = targetScope.canAdd();
+                var inRange = canRemove && canAdd;
+                var isCompatible = propertiesMatchAnyFieldset(properties, targetFieldsets);
```
Comment (kjac) [inline app/controllers/controller.js]: @Nicholas-Westby Performance suggestion: No need to traverse all the properties of all fieldsets if `inRange` is `false` here.

**Comment by kjac** [comment]
@Nicholas-Westby Great work! :+1: 

It looks very promising. I haven't tested it, tho', just glanced the code. 

A few things come to mind (bear with me if they're already handled):
- The source fieldset alias should match one of the target ones, since it's usually used for rendering (calling partials).
- Does this break the max allowed fieldsets feature?
- Can this entire feature be made opt-in through configuration?

File: app/views/archetype.default.html [hunk 2]
```diff
@@ -76,12 +86,8 @@
...
         </li>
     </ul>
-    <div ng-show="showAddButton()">
-        <a class="archetypeAddButton" ng-click="openFieldsetPicker(0, $event)" prevent-default>
-            <i class="icon icon-add"></i>
-            <archetype-localize key="addFieldset">Add an item</archetype-localize>
-        </a>
-    </div>
+
+    <!- Fieldset picker. -->
```
Comment (kgiszewski) [inline app/views/archetype.default.html]: Is this comment missing a dash in the opening part?

**Comment by kgiszewski** [comment]
@Nicholas-Westby This does look pretty sweet!

Aside from the questions @kjac mentioned, I'm not sure I have much feedback (other than this really adds flexibility #h5yr).

Regarding getting it into the next release. It's a matter of testing and figuring out if we can make this opt-in and can work through the max fieldsets potential issue.

**Comment by Nicholas-Westby** [comment]
Thanks for the comments. Some replies:

* I'll make sure to add the performance suggestion.
* I'll take a look to see if the fieldset alias is available on the model (don't see why it wouldn't be) to add an extra check.
* Nope, it respects the max fieldsets feature (see the "canAdd" variable).
* It can be made opt-in, though the trick would be figuring out which configuration. Since this deals with multiple configurations, I'm thinking the code will look at the drop target for that config (of course, users can then enable it for all their Archetypes). However, that will lead to funny situations like only being able to drag one way.
* Oops, I will fix that code comment (and now I know why GitHub was coloring that red).

**Comment by kgiszewski** [comment]
@Nicholas-Westby Good deal. I think the opt-in is just to make it explicit that you can now do this. Trying to avoid unintended consequences on existing installs.

**Comment by Nicholas-Westby** [comment]
I just noticed new fieldsets won't drag across boundaries. Seems like they are missing a few properties until they are saved. Here's a fieldset's property before being saved:

before

And here it is after being saved:

after

The before is missing the `dataTypeGuid`. I'm thinking I may need to add that to the model when it is first created (rather than on save). Any tips for going about that before I dig into it tonight?

**Comment by Nicholas-Westby** [comment]
Nevermind, think I figured out the issue with new fieldsets.

**Comment by Nicholas-Westby** [comment]
I think I've addressed all of your comments (see latest commits). Let me know if there are any further changes you'd like me to make.

**Comment by kgiszewski** [comment]
@Nicholas-Westby, I'm hoping to get to this to review this week.

**Comment by kgiszewski** [comment]
Yo, I'm reviewing this now. So far REALLY good :) 👍 

Only thing so far is a minor spacing thing I think we should consider:

screen shot 2016-04-28 at 8 52 59 am

**Comment by kgiszewski** [comment]
I'm still reviewing for any Gotchas™, but I'm really happy so far.

**Comment by kgiszewski** [comment]
Ok functionality review is complete. WOW. #h5yr!

**Comment by kgiszewski** [comment]
Looking through code now :)

**Comment by kgiszewski** [comment]
Looks like nice and clean code. @kjac, are you in a position to review this is in the next week or two?

**Comment by Nicholas-Westby** [comment]
Margin added :-)

**Comment by kgiszewski** [comment]
👍  I have some margin to fix in the config as well (unrelated to this PR).

**Comment by kjac** [comment]
I'll have a look at it tonight or tomorrow :-)

**Comment by kjac** [comment]
OK I've been fooling around with this a little while now and I really like the feature. However, there are a few issues with it.

- The `$scope.canSort()` method should probably take `$scope.canRemove()` into account. I was sorely confused until I realized I couldn't move my fieldset because it was the last one in the archetype, and `$scope.canRemove()` prevented the update.
- If I'm not mistaken, multiple properties can end up getting the same alias because of the way aliases are assigned to properties in `archetypeService.getUniquePropertyAlias()`. The simplest way to achieve this would be to drag a fieldset from a source archetype to a target archetype and then afterwards create a new fieldset of the same type in the source archetype (without saving first). Not sure if this is an issue, but I have a feeling it might just potentially be for editors relying on unique aliases.
- Validation is an issue. An invalid state is carried along from the source archetype to the target archetype, which looks funny. Even worse, if the target archetype does not have validation set up on the same properties, the validation breaks. This may be a configuration issue, but still worth addressing.
- The publishing feature is broken when dragging fieldsets between archetypes. This may be due to that merge issue @kgiszewski ran into, it looks like something isn't quite right (I'm testing on 7.4). I'll have to test some more on this.

A very positive note is that the RTE's aren't breaking. I had sort of expected that with all the grief we've had with them vs. fieldset sorting in the past.

**Comment by Nicholas-Westby** [comment]
Thanks for the thorough review! I'll try to address some of these tonight. Some comments:

* I'm not familiar with the scenario you are describing regarding `canSort()`. I'll see if I can figure it out, but if you could go into some more detail about this I would appreciate it.
* I'm not familiar with the "unique alias" stuff, so I'll dig into that to familiarize myself.
* I'll see if I can re-validate after the drop.
* Didn't see publishing issues myself, but I'll see if I can replicate and fix.
* Will see if ui-sortable has an option to indicate valid drop targets.

**Comment by kgiszewski** [comment]
Thanks guys, the next release will have some killer features. Thanks for your hard work.

**Comment by Nicholas-Westby** [comment]
Noticed something else I should probably look into, though I'm not certain it's particular to this new feature:

errors

Got those console errors when clicking the "reorder" button (or was it the "I am done reordering" button) in the document type designer:

reorder

Probably something related to the new 7.4 Umbraco overhaul.

**Comment by Nicholas-Westby** [comment]
Looks like even the current release of Archetype has issues with validations. I tried a few things and nothing has worked yet. Seems to be a problem of validations not being processed when fieldsets are collapsed.

Wasn't there some mention on another ticket recently about something relating to issues when fieldsets are collapsed? I think a solution may have been mentioned. Maybe it was to modify the `ng-if` that is removing fieldset contents when collapsed  so that it doesn't remove the contents if the fieldset has been opened once.

Maybe that would be worth trying. Let me know what you both think. In any event, it's getting late, so I'll have to pick this up again over the weekend.

**Comment by kjac** [comment]
@Nicholas-Westby The current release _has_ problems with validations of collapsed archetypes. Something that is partially mended by the upcoming release. The fix you propose is exactly the fix that has been implemented (or should have been, I still need to double check on that merge).

I think I need to clarify a bit on my findings from last night :)

- With regards to `canSort()`, you've changed that method to always return true. That's not really correct, as fieldsets cannot be moved out of an archetype, if the current limit for minimum fieldsets in that archetype is not fulfilled. I belive that if `canSort()` returns the value of `canRemove()`, this issue will be fixed. But it requires testing, of course :)
- The mentioned publishing issue has nothing to do with Umbraco publishing, but with the newly introduced publishing settings on fieldsets. You can enable publishing on fieldsets, which introduces publish and unpublish dates on the individual fieldsets. The problem is that the datepicker datatype doesn't save its value back to its model until the `formSubmitting` event is executed, so I had to add an `archetype-submit-watcher` to watch the publishing datepickers. Long story short, we need to make sure that these publishing dates are stored on submit, both for open and closed fieldsets (again, I need to double check that merge first and foremost).
- For ui-sortable there are a bunch of events you can hook into (see http://api.jqueryui.com/sortable/ - the `over` event might be the best option). If you can somehow render a drop target invalid, that'd of course be ideal. It that's not possible, maybe you can add a CSS class to the drop target placeholder that hides it or grays it out or gives some other indication that this is not a valid drop target.

**Comment by Nicholas-Westby** [comment]
I just merged the upstream master branch back into my forked feature branch. Perhaps that will help. I'm going to give it another go now.

**Comment by Nicholas-Westby** [comment]
I have the validation stuff playing mostly nicely on my local (no commits yet, as it is super hacky and needs some work). Basically removes the validations, then revalidates after the drop. Will try to get that finished up Sunday.

**Comment by Nicholas-Westby** [comment]
Strange, I thought I left a comment last night (perhaps I didn't press the "Comment" button).

Short version is the RTE dragging is borked and I'll need do so some more work on that (in addition to some other things). Validation's mostly done though.

**Comment by kgiszewski** [comment]
@Nicholas-Westby, ha. Welcome to the suck ;) RTE has long been a pain point ;) :hurtrealbad:

**Comment by Nicholas-Westby** [comment]
Looks like the RTE is an issue because the ID is generated with `archetypeService.getUniquePropertyAlias()`, which changes when dragging between fieldsets (that makes the disable/enable of the RTE fail). Perhaps I can tack on a data attribute or something to remember the original property ID to assist with the RTE fix. Will look into tomorrow.

**Comment by kjac** [comment]
@Nicholas-Westby it might be possible to replace `archetypeService.getUniquePropertyAlias()` with some usage of `_.uniqueId()` (see documentation: http://underscorejs.org/#uniqueId)

**Comment by Nicholas-Westby** [comment]
RTE's seem to be good and the validations are acting a little more sensible. I just clear the validation errors on the source/target Archetypes on drag drops. If errors still exist, they'll be made apparent during the next attempt to publish.

Will work on the further modifications tomorrow.

**Comment by kgiszewski** [comment]
@Nicholas-Westby 👍  great work, can't wait to test the final version

**Comment by Nicholas-Westby** [comment]
I think I've addressed all of the outstanding issues. Let me know if there's anything I missed.

**Comment by kgiszewski** [comment]
Super. Likely will review on Monday.

**Comment by kgiszewski** [comment]
@Nicholas-Westby, I haven't forgotten but the pesky day job is taking priority lately. I'll get to this soon :)

**Comment by Nicholas-Westby** [comment]
No worries. I can relate :-)

**Comment by kgiszewski** [comment]
@Nicholas-Westby I tried reviewing this today but got an error when trying to sort a fieldset. Core version 7.2.8.

screen shot 2016-05-11 at 1 17 47 pm

**Comment by Nicholas-Westby** [comment]
Strange, I haven't seen that on the latest version of Umbraco (7.4.3). I'll try building from my branch and installing it to a clean Umbraco 7.2.8 install tonight.

My best guess is that older versions of Umbraco use an older version of sortable before this bug was addressed: https://github.com/angular-ui/ui-sortable/issues/316

Probably related to this line of code: https://github.com/Nicholas-Westby/Archetype/blob/e1fa7a327194a463b414e862f4f9f0b11eb85c3c/app/controllers/controller.js#L76

Maybe there is some alternative approach to get the target sortable.

**Comment by kgiszewski** [comment]
Good info. The question will be what to do about it if we have to use the newer libs. Right now we only need core version 7.2.x. If we need 7.4.3, then we'll have to consider bumping up the dependency (not the best case). Let me know what you find out.

**Comment by Nicholas-Westby** [comment]
@kgiszewski FYI, my latest commit fixes this issue.

**Comment by kgiszewski** [comment]
@Nicholas-Westby looking real good. One issue remains that I'm debugging. When I disable the cross dragging, I can still do it in the UI.

**Comment by kgiszewski** [comment]
Disregard. Turns out it's just an interesting oddity that I think is possibly useful. So I have multiple Archetypes and I disabled cross-dragging on one and not on the other. The effect is you can 'give' one away and not let it come back. Any way... #h5yr! 👍

**Pull request merged**